### PR TITLE
time: use clock_gettime on macos

### PR DESCRIPTION
**Pull request by tamird**
Replace `gettimeofday` with `clock_gettime(CLOCK_REALTIME)` on:

```
all(target_os = "macos", not(target_arch = "aarch64")),
    target_os = "ios",
    target_os = "watchos",
    target_os = "tvos"
))]
```

`gettimeofday` was first used in
https://github.com/time-rs/time/commit/cc367edd953e72756ed6f0980918795c11e469b1
which predated the introduction of `clock_gettime` support in macOS
10.12 Sierra which became the minimum supported version in
58bbca958d917a89124da248735926f86c59a149.

Replace `mach_{absolute_time,timebase_info}` with
`clock_gettime(CLOCK_REALTIME)` on:

```
all(target_os = "macos", not(target_arch = "aarch64")),
    target_os = "ios",
    target_os = "watchos",
    target_os = "tvos"
))]
```

`mach_{absolute_time,timebase_info}` were first used in
https://github.com/time-rs/time/commit/cc367edd953e72756ed6f0980918795c11e469b1
which predated the introduction of `clock_gettime` support in macOS
10.12 Sierra which became the minimum supported version in
58bbca958d917a89124da248735926f86c59a149.

Note that this change was made for aarch64 in
5008a317ce8e508c390ed12bff281f307313376e which predated 10.12 becoming
the minimum supported version. The discussion took place in
https://github.com/rust-lang/rust/issues/91417 and in particular
https://github.com/rust-lang/rust/issues/91417#issuecomment-992151582
and
https://github.com/rust-lang/rust/issues/91417#issuecomment-1033048064
are relevant.

### CODE REVIEW

**Comment by tamird** [comment]
cc @thomcc and @hkratz who were involved in https://github.com/rust-lang/rust/issues/91417 and may generally be knowledgeable about this arcana.

File: library/std/src/sys/unix/time.rs [hunk 1]
```diff
@@ -216,209 +248,34 @@ impl From<__timespec64> for Timespec {
...
-    use crate::time::Duration;
-
-    use super::{SystemTime, Timespec};
-
-    #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-    pub struct Instant {
-        t: Timespec,
+impl Instant {
+    pub fn now() -> Instant {
+        Instant { t: Timespec::now(libc::CLOCK_MONOTONIC) }
```
Comment (thomcc) [inline library/std/src/sys/unix/time.rs:258]: This is a semantic change from the previous version (which used mach_absolute_time, which would be equivalent to CLOCK_UPTIME_RAW).

I would prefer we not make such a semantic change without consideration.

**Comment by thomcc** [comment]
I probably have more experience here than josh, so I'll take the review unless they object.

r? thomcc

File: library/std/src/sys/unix/time.rs (hunk 1 above)
Comment (tamird) [reply library/std/src/sys/unix/time.rs:258]: Yep, and we can keep the previous behavior. I called this out in the commit message, but I'll say it again here: I could not figure out from #91417 why the move to `CLOCK_UPTIME_RAW` was made.

Comment (thomcc) [reply library/std/src/sys/unix/time.rs:258]: It's because it's equivalent to `mach_absolute_time`.

**Comment by joshtriplett** [comment]
@thomcc By all means. I'm familiar with clock_gettime but I'm not a macOS expert at all.

**Comment by thomcc** [comment]
Yeah, I'm very familiar with the details of the macOS implementation of it.

File: library/std/src/sys/unix/time.rs (hunk 1 above)
Comment (thomcc) [reply library/std/src/sys/unix/time.rs:258]: To be more specific, there's a real way that `mach_absolute_time` is the true monotonic kernel timestamp. I used to think we should switch away from it, but so much of the kernel's internals (and even public APIs) are based around it that I think we should have a good reason to do so.

Comment (tamird) [reply library/std/src/sys/unix/time.rs:258]: Sounds good. I've updated this, have another look.

**Comment by tamird** [comment]
@thomcc any further comments?

**Comment by thomcc** [comment]
Sorry, I have to verify which clock the new apple guidance wants us to use. I have time this weekend to do it (I had hoped to do it this past weekend, but real life got in the way), but perhaps not before then.

**Comment by thomcc** [comment]
Indeed, it does not.

**Comment by tamird** [comment]
Friendly ping @thomcc - have you had a chance to think about this?

**Comment by tamird** [comment]
Hi @thomcc, are you still interested in reviewing this PR?

**Comment by thomcc** [comment]
Yeah this is fine for now.

@bors r+

**Pull request merged**